### PR TITLE
Implement "hidden" trigger for VisibilityV2 (IntersectionObserver based)

### PR DESCRIPTION
**Pull request by lannka**
For #5697

### CODE REVIEW

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 1]
```diff
@@ -421,6 +437,33 @@ export class Visibility {
   }
 
   /** @private */
+  onDocumentHidden_() {
+    for (let i = this.resources_.length - 1; i >= 0; i--) {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/visibility-impl.js]: I don't believe the backwards iteration is necessary here.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 2]
```diff
@@ -421,6 +437,33 @@ export class Visibility {
...
+      const resource = this.resources_[i];
+      if (!resource.hasLoadedOnce()) {
+        continue;
+      }
+
+      const listeners = this.listeners_[resource.getId()];
+      for (let j = listeners.length - 1; j >= 0; j--) {
+        const listener = listeners[j];
+        if (listener.shouldBeVisible) {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Can you explain what `shouldBeVisible` is?

Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: this indicates if it is a "visible" trigger or "hidden" trigger

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 3]
```diff
@@ -366,7 +367,15 @@ export class Visibility {
       this.resources_.push(resource);
     });
 
-    // TODO: support "hidden" spec.
+    // Hidden trigger
+    if (!this.visibilityListenerRegistered_) {
+      this.viewer_.onVisibilityChanged(() => {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Can we only register this when the first "hidden" visibility listener is registered?

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 1 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: changed

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 4]
```diff
@@ -402,7 +418,7 @@ export class Visibility {
 
           if (this.updateCounters_(
               lastChange.intersectionRatio * 100,
-              listener, shouldBeVisible)) {
+              listener, true)) {
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: `/* shouldBeVisible */ true`.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 5]
```diff
@@ -421,6 +437,33 @@ export class Visibility {
...
+      for (let j = listeners.length - 1; j >= 0; j--) {
+        const listener = listeners[j];
+        if (listener.shouldBeVisible) {
+          continue;
+        }
+
+        const state = listener.state;
+        const lastChange = state[LAST_CHANGE_ENTRY];
+        const lastVisible = lastChange ? lastChange.intersectionRatio * 100 : 0;
+        if (this.updateCounters_(lastVisible, listener, false)) {
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: ditto: Inline comment for `false`

**Comment by dvoytenko** [comment]
@lannka Same here: looks good, with a couple non-essential comments. However, would be nice to confirm with @avimehta

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 6]
```diff
@@ -382,7 +391,14 @@ export class Visibility {
       state[LAST_CHANGE_ENTRY] = change;
 
       // Update states and check if all conditions are satisfied
-      if (this.updateCounters_(visible, listener, shouldBeVisible)) {
+      const conditionsMet =
+          this.updateCounters_(visible, listener, shouldBeVisible);
+
+      if (!shouldBeVisible) {
+        // For "hidden" trigger, only update state, don't trigger.
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: The way I understand this, it looks like the hidden conditions will never trigger when the page is in background and the visibility conditions start meeting. Is that correct?

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 7]
```diff
@@ -382,7 +391,14 @@ export class Visibility {
       state[LAST_CHANGE_ENTRY] = change;
 
       // Update states and check if all conditions are satisfied
-      if (this.updateCounters_(visible, listener, shouldBeVisible)) {
+      const conditionsMet =
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: This is being used only once. WHy move it into a variable?

File: extensions/amp-analytics/0.1/test/test-visibility-impl.js [hunk 8]
```diff
@@ -638,6 +638,66 @@ describe('amp-analytics.visibility', () => {
       });
     });
 
+    it('"hidden" trigger should work with duration condition', () => {
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/test/test-visibility-impl.js]: add a test for hidden trigger without duration conditions?

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 7 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: because we want `this.updateCounters_` to run no matter of `shouldBeVisible`

File: extensions/amp-analytics/0.1/test/test-visibility-impl.js (hunk 8 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/test/test-visibility-impl.js]: Does it really make sense to have a hidden trigger without duration condition?

For example, what does it mean to have a spec like following?
```json
          "hidden": {
            "on": "hidden",
            "visibilitySpec": {
              "selector": "#img-2",
              "visiblePercentageMin": 10
            },
          }
```

To me, it should probably be done with a "visible" trigger instead.

Comment (avimehta) [inline extensions/amp-analytics/0.1/test/test-visibility-impl.js]: Sorry I meant without any conditions at all. so something like this:

```
          "hidden": {
            "on": "hidden",
            "visibilitySpec": {
              "selector": "#img-2"
            },
          }
```

A test like that might exist but was not sure and hence brought it up.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 6 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Yes, saw that. I think I might be understanding this wrong. So just want to make sure that I understand it correctly. What happens for following situation:

"hidden", minContinuousTime = 5 seconds

- Page is loaded, element is on screen for 2 seconds
- Page goes in background
- Does the trigger fire after 3 seconds or after the page comes to foreground again?

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 7 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: ooh. Missed that the if block was inside !shouldBeVisible. looks good.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 6 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: in your example, since the spec is about "continuous" time. It will fire after page back to foregound for another 5 seconds.

File: extensions/amp-analytics/0.1/test/test-visibility-impl.js (hunk 8 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/test/test-visibility-impl.js]: I'm not sure how exactly a spec like that should behave.

Does that mean "fires when page goes to background and the element has been visible"?
If so, I think a "visible" trigger should be used instead.

Or does that mean "fires when page goes to background and element is currently visible"? Why does people want to track that?

Comment (lannka) [inline extensions/amp-analytics/0.1/test/test-visibility-impl.js]: In fact, I think the V1 code for hidden trigger does not take non-duration condition into consideration either.

Comment (avimehta) [inline extensions/amp-analytics/0.1/test/test-visibility-impl.js]: Interesting. I thought my example would mean: "fires when page goes to background and element is currently visible". Anyway, I think this is another edge case that we should document.
